File: src/resolve/constraints.rs
//! Constraint types and parsing for dependency resolution.

use reqwest::blocking::Client;
use semver::VersionReq;

use crate::error::{AgentpackError, Result};
use crate::github::{list_tags, DEFAULT_GIT_REF};
use crate::manifest::{DepSpecToml, DepTable};

#[derive(Debug, Clone, Default)]
pub(super) struct ModuleConstraints {
    pub exact: Option<String>,
    pub branch: Option<String>,
    pub tag: Option<String>,
    pub semver_reqs: Vec<VersionReq>,
    /// Floating default branch
    pub latest: bool,
}

impl ModuleConstraints {
    pub fn merge(&mut self, other: ModuleConstraints) -> Result<()> {
        if let Some(e2) = other.exact {
            match &self.exact {
                Some(e1) if e1 != &e2 => {
                    return Err(AgentpackError::Cache(format!(
                        "conflicting commit pins for the same module: {e1} vs {e2}"
                    )));
                }
                None => self.exact = Some(e2),
                _ => {}
            }
        }
        if let Some(t2) = other.tag {
            match &self.tag {
                Some(t1) if t1 != &t2 => {
                    return Err(AgentpackError::Cache(format!(
                        "conflicting tags for the same module: {t1} vs {t2}"
                    )));
                }
                None => self.tag = Some(t2),
                _ => {}
            }
        }
        if let Some(b2) = other.branch {
            match &self.branch {
                Some(b1) if b1 != &b2 => {
                    return Err(AgentpackError::Cache(format!(
                        "conflicting branches for the same module: {b1} vs {b2}"
                    )));
                }
                None => self.branch = Some(b2),
                _ => {}
            }
        }
        self.semver_reqs.extend(other.semver_reqs);
        if other.latest {
            self.latest = true;
        }
        Ok(())
    }

    pub fn pick_git_ref(
        &self,
        client: &Client,
        owner: &str,
        repo: &str,
        force_refresh: bool,
    ) -> Result<String> {
        if let Some(c) = &self.exact {
            return Ok(c.clone());
        }
        if !self.semver_reqs.is_empty() {
            let tags = list_tags(client, owner, repo, force_refresh)?;
            let mut candidates: Vec<(semver::Version, String)> = Vec::new();
            for (name, _sha) in tags {
                let vpart = name.strip_prefix('v').unwrap_or(&name);
                if let Ok(v) = semver::Version::parse(vpart) {
                    candidates.push((v, name));
                }
            }
            candidates.sort_by(|a, b| b.0.cmp(&a.0));
            for (v, name) in candidates {
                if self.semver_reqs.iter().all(|r| r.matches(&v)) {
                    return Ok(name);
                }
            }
            return Err(AgentpackError::Cache(format!(
                "no tag matching semver constraints {:?} for {owner}/{repo}",
                self.semver_reqs
            )));
        }
        if let Some(t) = &self.tag {
            return Ok(t.clone());
        }
        if let Some(b) = &self.branch {
            return Ok(b.clone());
        }
        if self.latest {
            return Ok(DEFAULT_GIT_REF.into());
        }
        Ok(DEFAULT_GIT_REF.into())
    }
}

pub(super) fn from_ref_str(r: Option<&str>) -> Result<ModuleConstraints> {
    let Some(s) = r.map(str::trim).filter(|s| !s.is_empty()) else {
        return Ok(ModuleConstraints {
            latest: true,
            ..Default::default()
        });
    };
    if s.len() == 40 && s.chars().all(|c| c.is_ascii_hexdigit()) {
        return Ok(ModuleConstraints {
            exact: Some(s.to_lowercase()),
            ..Default::default()
        });
    }
    if let Ok(req) = VersionReq::parse(s) {
        return Ok(ModuleConstraints {
            semver_reqs: vec![req],
            ..Default::default()
        });
    }
    Ok(ModuleConstraints {
        tag: Some(s.to_string()),
        ..Default::default()
    })
}

pub(super) fn from_table(t: &DepTable, key_ref: Option<&str>) -> Result<ModuleConstraints> {
    if t.path.is_some() {
        return Err(AgentpackError::Cache(
            "path dependencies should be resolved before constraint parsing".into(),
        ));
    }
    let mut c = ModuleConstraints::default();
    let mut n = 0u8;
    if t.commit.is_some() {
        n += 1;
    }
    if t.tag.is_some() {
        n += 1;
    }
    if t.branch.is_some() {
        n += 1;
    }
    if t.version.is_some() {
        n += 1;
    }
    if n > 1 {
        return Err(AgentpackError::Cache(
            "dependency table may only specify one of commit, tag, branch, version".into(),
        ));
    }
    if let Some(commit) = &t.commit {
        c.exact = Some(commit.to_lowercase());
    } else if let Some(tag) = &t.tag {
        c.tag = Some(tag.clone());
    } else if let Some(branch) = &t.branch {
        c.branch = Some(branch.clone());
    } else if let Some(ver) = &t.version {
        c.semver_reqs.push(
            VersionReq::parse(ver).map_err(|e| AgentpackError::Cache(format!("semver: {e}")))?,
        );
    } else if let Some(r) = key_ref {
        c = from_ref_str(Some(r))?;
    } else {
        c.latest = true;
    }
    Ok(c)
}

pub(super) fn from_dep(dep: &DepSpecToml, key_ref: Option<&str>) -> Result<ModuleConstraints> {
    match dep {
        DepSpecToml::Short(s) => {
            let s = s.trim();
            if s.is_empty() {
                from_table(&DepTable::default(), key_ref)
            } else {
                from_ref_str(Some(s))
            }
        }
        DepSpecToml::Table(t) => from_table(t, key_ref),
    }
}

File: internal/resolve/constraints.go
package resolve

import (
	"encoding/hex"
	"fmt"
	"regexp"
	"sort"
	"strings"

	"github.com/Masterminds/semver/v3"

	githubsource "github.com/OlegHQ/agentpack/internal/github"
	"github.com/OlegHQ/agentpack/internal/manifest"
)

var bareVersionConstraint = regexp.MustCompile(`^[0-9]+(?:\.[0-9]+){0,2}(?:-[0-9A-Za-z.-]+)?(?:\+[0-9A-Za-z.-]+)?$`)

type versionRequirement struct {
	raw        string
	constraint *semver.Constraints
}

type ModuleConstraints struct {
	Exact  string
	Branch string
	Tag    string
	Semver []versionRequirement
	Latest bool
}

func (constraints *ModuleConstraints) Merge(other ModuleConstraints) error {
	if err := mergeUniquePin("commit", &constraints.Exact, other.Exact); err != nil {
		return err
	}
	if err := mergeUniquePin("tag", &constraints.Tag, other.Tag); err != nil {
		return err
	}
	if err := mergeUniquePin("branch", &constraints.Branch, other.Branch); err != nil {
		return err
	}
	constraints.Semver = append(constraints.Semver, other.Semver...)
	constraints.Latest = constraints.Latest || other.Latest
	return nil
}

type Tag struct {
	Name string
	SHA  string
}

type TagLister interface {
	ListTags(owner, repo string, forceRefresh bool) ([]Tag, error)
}

func (constraints ModuleConstraints) PickGitRef(lister TagLister, owner, repo string, forceRefresh bool) (string, error) {
	if constraints.Exact != "" {
		return constraints.Exact, nil
	}
	if len(constraints.Semver) != 0 {
		tags, err := lister.ListTags(owner, repo, forceRefresh)
		if err != nil {
			return "", fmt.Errorf("list tags for %s/%s: %w", owner, repo, err)
		}
		type candidate struct {
			version *semver.Version
			name    string
		}
		candidates := make([]candidate, 0, len(tags))
		for _, tag := range tags {
			versionText := strings.TrimPrefix(tag.Name, "v")
			version, err := semver.StrictNewVersion(versionText)
			if err == nil {
				candidates = append(candidates, candidate{version: version, name: tag.Name})
			}
		}
		sort.Slice(candidates, func(i, j int) bool { return candidates[i].version.GreaterThan(candidates[j].version) })
		for _, candidate := range candidates {
			matched := true
			for _, requirement := range constraints.Semver {
				if !requirement.constraint.Check(candidate.version) {
					matched = false
					break
				}
			}
			if matched {
				return candidate.name, nil
			}
		}
		raw := make([]string, len(constraints.Semver))
		for index, requirement := range constraints.Semver {
			raw[index] = requirement.raw
		}
		return "", fmt.Errorf("no tag matching semver constraints %v for %s/%s", raw, owner, repo)
	}
	if constraints.Tag != "" {
		return constraints.Tag, nil
	}
	if constraints.Branch != "" {
		return constraints.Branch, nil
	}
	return githubsource.DefaultGitRef, nil
}

func ConstraintsFromRef(value string, present bool) (ModuleConstraints, error) {
	value = strings.TrimSpace(value)
	if !present || value == "" {
		return ModuleConstraints{Latest: true}, nil
	}
	if isCommitSHA(value) {
		return ModuleConstraints{Exact: strings.ToLower(value)}, nil
	}
	if requirement, err := parseCargoRequirement(value); err == nil {
		return ModuleConstraints{Semver: []versionRequirement{requirement}}, nil
	}
	return ModuleConstraints{Tag: value}, nil
}

func ConstraintsFromDependency(dependency manifest.Dependency, keyRef string, hasKeyRef bool) (ModuleConstraints, error) {
	if dependency.Short != nil {
		value := strings.TrimSpace(*dependency.Short)
		if value != "" {
			return ConstraintsFromRef(value, true)
		}
		return ConstraintsFromTable(manifest.DependencyTable{}, keyRef, hasKeyRef)
	}
	if dependency.Table == nil {
		return ModuleConstraints{}, fmt.Errorf("dependency has neither string nor table value")
	}
	return ConstraintsFromTable(*dependency.Table, keyRef, hasKeyRef)
}

func ConstraintsFromTable(table manifest.DependencyTable, keyRef string, hasKeyRef bool) (ModuleConstraints, error) {
	if table.Path != nil {
		return ModuleConstraints{}, fmt.Errorf("path dependencies should be resolved before constraint parsing")
	}
	count := 0
	for _, present := range []bool{table.Commit != nil, table.Tag != nil, table.Branch != nil, table.Version != nil} {
		if present {
			count++
		}
	}
	if count > 1 {
		return ModuleConstraints{}, fmt.Errorf("dependency table may only specify one of commit, tag, branch, version")
	}
	switch {
	case table.Commit != nil:
		return ModuleConstraints{Exact: strings.ToLower(*table.Commit)}, nil
	case table.Tag != nil:
		return ModuleConstraints{Tag: *table.Tag}, nil
	case table.Branch != nil:
		return ModuleConstraints{Branch: *table.Branch}, nil
	case table.Version != nil:
		requirement, err := parseCargoRequirement(*table.Version)
		if err != nil {
			return ModuleConstraints{}, fmt.Errorf("semver: %w", err)
		}
		return ModuleConstraints{Semver: []versionRequirement{requirement}}, nil
	case hasKeyRef:
		return ConstraintsFromRef(keyRef, true)
	default:
		return ModuleConstraints{Latest: true}, nil
	}
}

func parseCargoRequirement(value string) (versionRequirement, error) {
	value = strings.TrimSpace(value)
	if value == "" || strings.HasPrefix(value, "v") {
		return versionRequirement{}, fmt.Errorf("invalid semantic version requirement %q", value)
	}
	translated := value
	if bareVersionConstraint.MatchString(value) {
		translated = "^" + value
	}
	constraint, err := semver.NewConstraint(translated)
	if err != nil {
		return versionRequirement{}, err
	}
	return versionRequirement{raw: value, constraint: constraint}, nil
}

func mergeUniquePin(kind string, target *string, incoming string) error {
	if incoming == "" {
		return nil
	}
	if *target != "" && *target != incoming {
		plural := kind + "s"
		if kind == "branch" {
			plural = "branches"
		} else if kind == "commit" {
			plural = "commit pins"
		}
		return fmt.Errorf("conflicting %s for the same module: %s vs %s", plural, *target, incoming)
	}
	if *target == "" {
		*target = incoming
	}
	return nil
}

func isCommitSHA(value string) bool {
	if len(value) != 40 {
		return false
	}
	_, err := hex.DecodeString(value)
	return err == nil
}
